lib/src/matrix/decomposition/singular_value.dart

```diff
@@ -1,483 +1,951 @@
+              // A part of column "l" of matrix VT from row 0 to end multiply by -1
+              for (i = 0; i < columnsA; i++) {
+                v[(l * columnsA) + i] = v[(l * columnsA) + i] * -1.0;
               }
             }
+          }
 
-            // Order the singular values.
-            while (k < pp) {
-              if (_s[k] >= _s[k + 1]) {
-                break;
-              }
-              var t = _s[k];
-              _s[k] = _s[k + 1];
-              _s[k + 1] = t;
-              if (k < _n - 1) {
-                for (var i = 0; i < _n; i++) {
-                  t = _v.getUnchecked(i, k + 1);
-                  _v.setUnchecked(i, k + 1, _v.getUnchecked(i, k));
-                  _v.setUnchecked(i, k, t);
-                }
+          // Order the singular value.
+          while (l != mn - 1) {
+            if (stemp[l] >= stemp[l + 1]) {
+              break;
+            }
+
+            t = stemp[l];
+            stemp[l] = stemp[l + 1];
+            stemp[l + 1] = t;
+            if (computeVectors && l < columnsA) {
+              // Swap columns l, l + 1
+              for (i = 0; i < columnsA; i++) {
+                var a = v[((l + 1) * columnsA) + i];
+                var b = v[(l * columnsA) + i];
+                v[(l * columnsA) + i] = a;
+                v[((l + 1) * columnsA) + i] = b;
               }
-              if (k < _m - 1) {
-                for (var i = 0; i < _m; i++) {
-                  t = _u.getUnchecked(i, k + 1);
-                  _u.setUnchecked(i, k + 1, _u.getUnchecked(i, k));
-                  _u.setUnchecked(i, k, t);
-                }
+            }
+
+            if (computeVectors && l < rowsA) {
+              // Swap columns l, l + 1
+              for (i = 0; i < rowsA; i++) {
+                var a = u[((l + 1) * rowsA) + i];
+                var b = u[(l * rowsA) + i];
+                u[(l * rowsA) + i] = a;
+                u[((l + 1) * rowsA) + i] = b;
               }
-              k++;
             }
-            iter = 0;
-            p--;
+
+            l = l + 1;
           }
+
+          iter = 0;
+          m = m - 1;
           break;
       }
     }
+
+    if (computeVectors) {
+      // Finally transpose "v" to get "vt" matrix
+      for (i = 0; i < columnsA; i++) {
+        for (j = 0; j < columnsA; j++) {
+          vt[(j * columnsA) + i] = v[(i * columnsA) + j];
+        }
+      }
+    }
+
+    // Copy stemp to s with size adjustment. We are using ported copy of linpack's svd code and it uses
+    // a singular vector of length rows+1 when rows < columns. The last element is not used and needs to be removed.
+    // We should port lapack's svd routine to remove this problem.
+    for (var i = 0; i < min(rowsA, columnsA); i++) {
+      s[i] = stemp[i];
+    }
   }
 
-  /// internal storage of U and V.
-  final Matrix<double> _u, _v;
+  /// Solves a system of linear equations, AX = B, with A SVD factorized.
+  dynamic solve(dynamic input) {
```

Comment:
Please avoid `dynamic` types, neither argument nor result should be `null`. Probably best to create two methods, `solveMatrix` and `solveVector`. 

Possibly you can add a dispatcher named `solve` following the pattern here, but not convinced that would work well with the changed return type: https://github.com/renggli/dart-data/blob/e7c5ab69bf05393ba12f769cf9af45f2eb3c1e92/lib/src/matrix/operator/mul.dart#LL10C67-L10C67

Reply:
Updated

````dart
  Matrix<double> solve(Object input) {
     ...
    if (input is Matrix<num>) {
    ...
    } else if (input is Vector<num>) {
   ...
    } else {
      throw ArgumentError('Not supported input.');
    }
  }
````